core/src/server/rpc_module.rs

```diff
@@ -762,34 +702,42 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 		let callback = {
 			self.methods.verify_and_insert(
 				subscribe_method_name,
-				MethodCallback::new_subscription(Arc::new(move |id, params, method_sink, conn, claimed| {
+				MethodCallback::new_subscription(Arc::new(move |id, params, method_sink, conn| {
 					let uniq_sub = SubscriptionKey { conn_id: conn.conn_id, sub_id: conn.id_provider.next_id() };
 
 					// response to the subscription call.
 					let (tx, rx) = oneshot::channel();
 
-					let sink = SubscriptionSink {
+					let sink = PendingSubscriptionSink {
 						inner: method_sink,
-						close_notify: Some(conn.close_notify),
 						method: notif_method_name,
 						subscribers: subscribers.clone(),
 						uniq_sub,
-						id: Some((id.clone().into_owned(), tx)),
-						unsubscribe: None,
-						_claimed: claimed,
+						id: id.clone().into_owned(),
+						subscribe: tx,
 					};
 
-					// The callback returns a `SubscriptionResult` for better ergonomics and is not propagated further.
-					if callback(params, sink, ctx.clone()).is_err() {
-						tracing::warn!("Subscribe call `{}` failed", subscribe_method_name);
-					}
+					// The subscription callback is a future from the subscription
+					// definition and not the as same when the subscription call has been completed.
+					//
+					// This runs until the subscription callback has completed.
+					let sub_fut = callback(params.into_owned(), sink, ctx.clone());
+
+					tokio::spawn(async move {
+						if sub_fut.await.is_err() {
+							tracing::warn!("Subscribe call `{subscribe_method_name}` closed");
```

Comment:
Is it worth logging the error that caused it to close here?

Reply:
we could but currently the error cause is erased by the `SubscriptionEmptyError`.

so there is no point at the moment

Reply:
Did we used to do that because we didn't return an `async` thing and the fn would return probably before the subscription has ended?

Does it make any sense now to allow the subscription to return a normal error object or whatever, and to send that back to the client before closing the connection if possible?

I wonder whether doing this would keep subscriptions easy to understand; if they return an error, the error will be sent to the user on close. If they return `Ok(())`, no explicit error will be sent. Perhaps we don't need an explicit `sink.close(msg)` type thing in that case; it's closed when the async block returns?

Just mulling over ideas here since right now this Ok/Err thing is sortof just ignored.

Reply:
> Did we used to do that because we didn't return an async thing and the fn would return probably before the subscription has ended?

Returning an error is a hack for ergonomics i.e, to bail easily from the user defined, it's not really used except logging it here. We could remove it...

> Does it make any sense now to allow the subscription to return a normal error object or whatever, and to send that back to the client before closing the connection if possible?

Not really, because we have this `accept/reject` API which handles the JSON-RPC response to the call.
If in an error happens after that it's simply wrong to send out it as an JSON-RPC response. That's why we have a oneshot channel which indicates whether the "subscription method call" was answered or not. If it's not answered then a default JSON-RPC error response is sent out.